homeassistant/components/forked_daapd/media_player.py

```diff
@@ -0,0 +1,669 @@
+            if self.volume_level == 0:
+                return
+            self._last_volume = self.volume_level  # store volume level to restore later
+            target_volume = 0
+        else:
+            target_volume = self._last_volume  # restore volume level
+        await self.async_set_volume_level(volume=target_volume)
+
+    async def async_set_volume_level(self, volume):
+        """Set volume - input range [0,1]."""
+        await self._api.set_volume(volume=volume * 100, output_id=self._output_id)
+
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        return SUPPORTED_FEATURES_ZONE
+
+
+class ForkedDaapdMaster(MediaPlayerDevice):
+    """Representation of the main forked-daapd device."""
+
+    def __init__(
+        self, clientsession, name, ip_address, api_port, api_password, default_volume
+    ):
+        """Initialize the ForkedDaapd Master Device."""
+        self._api = ForkedDaapdAPI(clientsession, ip_address, api_port, api_password)
+        self._player = STARTUP_DATA[
+            "player"
+        ]  # _player, _outputs, and _queue are loaded straight from api
+        self._outputs = STARTUP_DATA["outputs"]
+        self._queue = STARTUP_DATA["queue"]
+        self._updater = None  # _updater is instance of a helper class which takes care of data and entity updates
+        self._track_info = defaultdict(
+            str
+        )  # _track info is found by matching _player data with _queue data
+        self._last_outputs = None  # used for device on/off
+        self._default_volume = default_volume  # used for mute/unmute
+        self._last_volume = None
+        self._player_last_updated = None
+        self._pipe_control_api = None
+        self._name = name
+        self._ip_address = ip_address
+        self._api_port = api_port
+        self._api_password = api_password
+        self._tts_pause_time = DEFAULT_TTS_PAUSE_TIME
+        self._tts_volume = DEFAULT_TTS_VOLUME
+        self._tts_requested = False
+        self._tts_queued = False
+        self._tts_playing_event = asyncio.Event()
+        self._on_remove = None
+        self._clientsession = clientsession
+
+    async def async_added_to_hass(self):
+        """Use lifecycle hooks."""
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, SIGNAL_UPDATE_PLAYER, self._update_player
+            )
+        )
+        self.async_on_remove(
+            async_dispatcher_connect(self.hass, SIGNAL_UPDATE_QUEUE, self._update_queue)
+        )
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, SIGNAL_UPDATE_OUTPUTS, self._update_outputs
+            )
+        )
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, SIGNAL_UPDATE_MASTER, self._update_callback
+            )
+        )
+        self._updater = ForkedDaapdUpdater(self.hass, self._api)
```

Comment:
We should move the updater outside the entity and store the updater in hass.data if needed. Create the `api` instance first and then pass that to both the master entity and the updater when instantiating those.

Reply:
Use signals to tie updater to master lifecycle?

Reply:
Yes. :+1: